src/peft/tuners/osf/model.py

```diff
@@ -0,0 +1,192 @@
+
+    @staticmethod
+    def _check_target_module_exists(osf_config, key):
+        return check_target_module_exists(osf_config, key)
+
+    def _mark_only_adapters_as_trainable(self, model: nn.Module) -> None:
+        for n, p in model.named_parameters():
+            if (
+                self.prefix not in n
+                and "svd_params" not in n
+                and not n.endswith(("_U_low", "_S_low", "_V_low"))
+            ):
+                p.requires_grad = False
+
+    def _set_adapter_layers(self, enabled: bool = True) -> None:
+        pass
+
+    def enable_adapter_layers(self) -> None:
+        self._set_adapter_layers(True)
+
+    def disable_adapter_layers(self) -> None:
+        self._set_adapter_layers(False)
+
+    def set_adapter(self, adapter_name):
+        self.active_adapter = adapter_name
+
+    def _cast_adapter_dtype(self, adapter_name: str, autocast_adapter_dtype: bool = True) -> None:
+        """
+        Ensure all OSF adapter components have consistent dtype with the base model.
+
+        Instead of forcing float32, we match the base model's actual dtype for consistency.
+        """
+        if not autocast_adapter_dtype:
+            return
+
+        for module in self.model.modules():
+            if not hasattr(module, 'osf_svd_params'):
+                continue
+
+            # Get target dtype from base layer weight
+            base_layer = getattr(module, 'base_layer', None)
+            if base_layer is None or not hasattr(base_layer, 'weight'):
+                continue
+
+            target_dtype = base_layer.weight.dtype
+
+            # Cast trainable low-rank parameters to match base model dtype
+            if adapter_name in module.osf_svd_params:
+                svd_params = module.osf_svd_params[adapter_name]
+                for param_name, param in svd_params.items():
+                    if param.dtype != target_dtype:
+                        param.data = param.data.to(target_dtype)
+
+            # Cast frozen high-rank buffers to match base model dtype
+            for buffer_dict_name in OSFLayer.other_param_names:
+                if hasattr(module, buffer_dict_name):
+                    buffer_dict = getattr(module, buffer_dict_name)
+                    if adapter_name in buffer_dict:
+                        buffer = buffer_dict[adapter_name]
+                        if buffer.dtype != target_dtype:
+                            buffer_dict[adapter_name] = buffer.to(target_dtype)
+
+    def unload(self):
+        raise NotImplementedError("OSF models cannot be unloaded yet")
+
+    def merge_adapter(self, *args, **kwargs):
+        raise NotImplementedError("OSF models do not support merging")
+
+    def unmerge_adapter(self, *args, **kwargs):
+        raise NotImplementedError("OSF models do not support merging")
+
+    def merge_and_unload(self, *args, **kwargs):
+        raise NotImplementedError("OSF models do not support merging")
```

Comment:
```suggestion
    def unmerge_adapter(self, *args, **kwargs):
        raise NotImplementedError("OSF models do not support merging")
```

IIUC only `unmerge` is unsupported. Since #2771 is now merged we can remove these implementations and use the `BaseTuner` implementation which calls `layer.merge` under the hood.

Reply:
This makes sense made the following changes:

- Removed OSF specific overrides for merge_adapter and merge_and_unload to use BaseTuner's implementations.
- Kept unmerge unsupported to raise NotImplementedError.
- merge and merge_and_unload work via BaseTuner and unmerge remains explicitly disallowed for OSF.